index.html

```diff
@@ -164,6 +164,281 @@ <h2>
+              <code>web_push</code> (required)
+            </dt>
+            <dd>
+              <p>
+                An integer that must be 9001. Used to disambiguate a [=declarative push message=]
+                from other JSON documents.
+              </p>
+            </dd>
+            <dt>
+              <code>notification</code> (required)
+            </dt>
+            <dd>
+              <p>
+                A JSON object consisting of the following members, many being analogous to
+                Notifications API features.
+              </p>
+              <dl>
+                <dt>
+                  <code>title</code> (required)
+                </dt>
+                <dd>
+                  <p>
+                    A string.
+                  </p>
+                </dd>
+                <dt>
+                  <code>dir</code>
+                </dt>
+                <dd>
+                  <p>
+                    "<code>auto</code>", "<code>ltr</code>", or "<code>rtl</code>".
+                  </p>
+                </dd>
+                <dt>
+                  <code>lang</code>
+                </dt>
+                <dd>
+                  <p>
+                    A string that holds a language tag.
+                  </p>
+                </dd>
+                <dt>
+                  <code>body</code>
+                </dt>
+                <dd>
+                  <p>
+                    A string.
+                  </p>
+                </dd>
+                <dt>
+                  <code>url</code> (required)
+                </dt>
+                <dd>
+                  <p>
+                    A string that holds a URL.
+                  </p>
+                </dd>
+                <dt>
+                  <code>tag</code>
+                </dt>
+                <dd>
+                  <p>
+                    A string.
+                  </p>
+                </dd>
+                <dt>
+                  <code>icon</code>
+                </dt>
+                <dd>
+                  <p>
+                    A string that holds a URL.
+                  </p>
+                </dd>
+                <dt>
+                  <code>silent</code>
+                </dt>
+                <dd>
+                  <p>
+                    A boolean.
+                  </p>
+                </dd>
+                <dt>
+                  <code>data</code>
+                </dt>
+                <dd>
+                  <p>
+                    Any JSON value.
+                  </p>
+                </dd>
+                <dt>
+                  <code>mutable</code>
+                </dt>
+                <dd>
+                  <p>
+                    A boolean. When true causes a <code>pushnotification</code> to be dispatched to
+                    a service worker (if any).
+                  </p>
+                </dd>
+              </dl>
+            </dd>
+            <dt>
+              <code>app_badge</code>
```

Comment:
There really are three features being added here:
1. Declarative push notifications (w/ the Web Notification API),
2. Declarative push notifications that also fire a JavaScript event,
3. Declarative app badge updating (w/ the Badging API).

I think it'd be helpful to think about them separately. Intuitively I'm in favour of (1) and (3), however I have some concerns about (2).

A large part of our interest in declarative notifications comes from device resource usage improvements that it enables in Web Push message delivery, and from time-to-feedback improvements when users activate a Web Notification.

If the design were to stick to features (1) and (3), it would remove the need to launch a renderer process in Chrome entirely, which brings major savings (& with enough time and determination we could even trim much of our browser process). The `mutable` property would remove that optimisation, and based on our experience I expect it to be adopted for purposes of (at least) sending analytics pings back to the server. There may be other ways to solve for that use case too.

@beidson describes this as:
> We also propose a more flexible event handling model that developers can opt into to transform a push notification. This new model still depends on Service Workers, but provides the privacy-preserving guarantee that there is always a fallback notification to be shown, so pushes cannot trigger silent background runtime.

Existing implementations already solve for this: Chrome shows a default notification on behalf of the website, and IIRC Firefox and Edge penalize and eventually remove the underlying push subscription. I don't understand what benefit this brings over existing functionality.

I propose scoping out this concept from the proposal as it'll need more discussion.

Reply:
@beverloo in case you missed this, I split out this discussion into #391. We consider the mutable field a very important part of the overall value proposition as without it web developers cannot uniformly adopt this more robust delivery mechanism.